### PR TITLE
feat: support for virtual components

### PR DESCRIPTION
**Pull request by habansal**
### Main changes of this PR are -
  - the new API - `withMappable `
  - Updating `get editProps()` in `EditableComponent.tsx`

### Description of the changes
* Added a new `withMappable` Utility to make a React component mappable and add the edition capabilities and model data
  -  basically the aggregation of three utilities to simplify use by FED - `withEditorContext(withModel(withEditable...`
  -  refactored `ComponentMapping.map` to use that as well
  -  See how it will be used in the sample app - https://github.com/sharanyavinod/spa-sample-app/blob/develop/src/components/Text.js

* Adds a new `data-cq-resource-type` attribute on the component editable dom, when passed via the EditConfig
  -  Note that this attribute will always be present and  the responsibility to check whether a component is virtual or not will be (and should be) of the back-end

* Added unit tests for the fix

* Fixed linting issues.

### CODE REVIEW

File: src/ComponentWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Wraps a React component with the edition capabilities and model data
+ * If an {@link EditConfig} is provided the <i>clazz</i> is wrapped to provide edition capabilities on the AEM Page Editor
+ *
+ * @param {React.Component} component - class to be instantiated for the given resource types
+ * @param {EditConfig} [editConfig] - configuration object for enabling the edition capabilities
+ * @param {ReloadableModelProperties} modelConfig - configuration object of the withModel function.
+ * @returns {any}
+ */
+export const WrapAsAEMEditableComponent = <P extends MappedComponentProperties>(
```
Comment (pfauchere) [inline src/ComponentWrapper.tsx:28]: what about `withEditablePlaceholder`?

File: src/components/EditableComponent.tsx [hunk 2]
```diff
@@ -19,12 +19,14 @@ import { ContainerState } from './Container';
...
- * @property emptyLabel Label to be displayed on the overlay when the component is empty.
- * @property isEmpty Callback function to determine if the component is empty.
+ * @property emptyLabel - Label to be displayed on the overlay when the component is empty
+ * @property isEmpty - Callback function to determine if the component is empty
+ * @property aemResourceType - AEM ResourceType to be added as an attribute on the editable component dom
  */
 export interface EditConfig<P extends MappedComponentProperties> {
     emptyLabel?: string;
     isEmpty(props: P): boolean;
+    aemResourceType?: string;
```
Comment (pfauchere) [inline src/components/EditableComponent.tsx:29]: As discussed `resourceType` should be enough as the object is already meant for configuring the editability of the component in AEM

File: .gitignore [hunk 3]
```diff
@@ -3,3 +3,4 @@ dist/
 .scannerwork/
 **/*.log
 *.tgz
+.idea
```
Comment (CezCz) [inline .gitignore:6]: Please put this in your global .gitignore in your home directory not here.

File: src/ComponentWrapper.tsx (hunk 1 above)
Comment (sharanyavinod) [reply src/ComponentWrapper.tsx:28]: Wouldn't `withPlaceholder` signify that it mostly just adds a placeholder to the component when this wrapper also pulls the data from the model and sets editable layers? `Placeholder` seems apt for virtual cases whereas this wrapper could ideally be used for all standalone components which are not using `MapTo`. WDYT @pfauchere @habansal ?

Comment (pfauchere) [reply src/ComponentWrapper.tsx:28]: @sharanyavinod, I get your point. The scope is broader. Do you have suggestions?

Comment (habansal) [reply src/ComponentWrapper.tsx]: As discussed, we are changing to a common utility - `withMappable`  and doing some refactorig

File: src/components/ModelProvider.tsx [hunk 4]
```diff
@@ -74,15 +75,21 @@ export class ModelProvider extends Component<ModelProviderType> {
...
         ModelManager.getData({ path, forceReload: this.props.cqForceReload}).then((data: Model) => {
           if (data && (Object.keys(data).length > 0)) {
               const props = Utils.modelToProps(data);
+
               this.setState(props);
           }
+        }).catch((error) => {
+            if (error.response.status === 404) {
+                this.setState({ 'virtual': true });
+            }
```
[the remark this answers is not in the document]
Comment (habansal) [reply src/components/ModelProvider.tsx]: As discussed, updated the PR to always have `data-cq-resource-type` attribute when passed via the Editconfig

**Review by CezCz — APPROVED** [review]

**Pull request merged**